### DISCUSSION TITLE
Removing specific messages from archive

### DISCUSSION THREAD

**Discussion by kawaii44** [processone/ejabberd#3536] [Questions] [answered]
I have an ejabberd service using an external PostgreSql DB and I want to be able to remove selected archived messages posted by certain users into group chats using something like
 
`remove_mam_from_user_with_peer  xyz@conference.example.com example.com user@example.com`
 
I can use ejabberdctl, but prefer using the REST Admin API
 
I have tried many ways,  but get internal errors (and Http 400 – bad request errors) which I think are related to using the ‘@’ in the identifier.
 
Is there a way to get this admin command to work?
If not, is there another command that I can use?
If not, will I do harm if I use SQL to delete the messages directly from the external PostgreSQL mam database?
 
Thank you

**Comment by badlop** [comment, marked as answer]
Umm, if I perform a call like you do, no error is returned, but no messages are removed. So, it doesn't work correctly.

```
$ ejabberdctl remove_mam_for_user_with_peer sala1@conference.localhost localhost user1@localhost
MAM archive removed

$ echo $?
0
```
Looking at the command help, you are not providing the arguments correctly. The _user_ argument should be only the username (in this case the roomname) See https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#remove-mam-for-user-with-peer

But, when I execute the command with the correct arguments, it crashes:

```
$ ejabberdctl remove_mam_for_user_with_peer room1 conference.localhost user1@localhost
Unhandled exception occurred executing the command:
** exception error: {module_not_loaded,mod_mam,<<"conference.localhost">>}
   in function  gen_mod:get_module_opts/2 (src/gen_mod.erl, line 338)
   in call from gen_mod:get_module_opt/3 (src/gen_mod.erl, line 318)
   in call from gen_mod:db_mod/3 (src/gen_mod.erl, line 352)
   in call from mod_mam:remove_mam_for_user_with_peer/3 (/home/bernar/e/git/ejabberd/src/mod_mam.erl, line 329)
   in call from ejabberd_ctl:call_command/4 (src/ejabberd_ctl.erl, line 315)
   in call from ejabberd_ctl:try_call_command/4 (src/ejabberd_ctl.erl, line 280)
   in call from ejabberd_ctl:process2/4 (src/ejabberd_ctl.erl, line 233)
   in call from ejabberd_ctl:process/2 (src/ejabberd_ctl.erl, line 211)
```

So, there's a bug that breaks this command when removing MAM archives instead of user->user archives.

I have to apply this patch for the command to work correctly. Can you try it?
```diff
diff --git a/src/mod_mam.erl b/src/mod_mam.erl
index 4641836f6..12542bfa5 100644
--- a/src/mod_mam.erl
+++ b/src/mod_mam.erl
@@ -326,7 +326,7 @@ remove_mam_for_user_with_peer(User, Server, Peer) ->
     LServer = jid:nameprep(Server),
     try jid:decode(Peer) of
 	Jid ->
-	    Mod = gen_mod:db_mod(LServer, ?MODULE),
+	    Mod = get_module_host(LServer),
 	    case Mod:remove_from_archive(LUser, LServer, Jid) of
 		ok ->
 		    {ok, <<"MAM archive removed">>};
@@ -339,6 +339,12 @@ remove_mam_for_user_with_peer(User, Server, Peer) ->
 	{error, <<"Invalid peer JID">>}
     end.
 
+get_module_host(LServer) ->
+    try gen_mod:db_mod(LServer, ?MODULE)
+    catch error:{module_not_loaded, ?MODULE, LServer} ->
+        gen_mod:db_mod(ejabberd_router:host_of_route(LServer), ?MODULE)
+    end.
+
 -spec get_room_config([muc_roomconfig:property()], mod_muc_room:state(),
 		      jid(), binary()) -> [muc_roomconfig:property()].
 get_room_config(Fields, RoomState, _From, _Lang) ->
```

**Reply by mremond** [reply]
@kawaii44 @badlop What's thew update ?
Should we apply this patch ?

**Reply by badlop** [reply]
That small patch has been applied in https://github.com/processone/ejabberd/commit/31884f6c9d8686ad2c15b04c8c585e23e37be0c2
